Reimplement PR 112032, preserve trailing extra data when updating ZIP files

edwardneal opened the pull request:
Contributes to #112017. Reimplements #112032. From the PR description:

> aapt2 creates ZIP entries with a "version needed to extract" of 0.0, but sometimes writes these entries with deflate compression (which requires a "version needed to extract" of 2.0.)
>
> When a ZipArchive is opened in Update mode and a new item is added, the central directory is written with the correct value for this field. However, the local file header for existing entries isn't rewritten (because it hasn't changed.) This leads to a mismatch between the local file header values and the central directory header values, which causes Android app builds to fail.
>
> When an existing entry's "version needed to extract" is changed, we force its local file header to be rewritten.
>
> There's a test to cover this. I've also added an entry to the packaged_resources.zip file from the issue and confirmed that the field matches between the local file header and the CD header for all compressed entries in the resultant file.

The second commit enables ZipArchive to perform in-place updates of ZIP entries with slightly malformed "extra data" fields. This field must consist of [header, data] pairs, where the header is four bytes long. This isn't always adhered to though - zipalign will pad the extra data field with zeroes to make sure that the compressed data which follows is aligned to an X byte boundary.

To tolerate this, we introduce the concept of "trailing extra data" - the part of the extra data field which can't be successfully parsed as a [header, data] pair. This is detected when parsing the central directory header (and the local file header), used when calculating the "extra data length" part of the relevant header, then written out after the parsed extra data.

edwardneal committed:
Reimplement PR 112032

edwardneal committed:
Permit trailing malformed extra field data

This data is generated by zipalign. It needs to be accounted for and preserved.

carlossanlop commented:
Feel free to mark it as ready to review when you're ready and ping me to give it a full review.

edwardneal commented:
Thanks @carlossanlop - I can see enough tests passing on enough platforms for review.

Besides the tests, I've rolled the latest build of System.IO.Compression into a local .NET 10 preview installation and published a simple .NET for Android project. I was able to reproduce the problem resolved by PR 112032 with this, but not the issue with ZIP alignment. To reproduce that, I ran `zipalign` against a ZIP file and tried to modify it.

ericstj commented:
@carlossanlop  -- seems like https://github.com/dotnet/runtime/commit/44c08b874792487f81c546ea112b7d4a677c911e created some conflicts here -- can you work with @edwardneal to get those resolved and review this change, please?

edwardneal committed:
Merged main

edwardneal commented:
Thanks both. I've merged and verified that all's well in CI, and have managed to successfully build an Android .apk and .aab file using the new version. This is ready to review.

carlossanlop committed:
Merge branch 'main' into bug/issue-112017-ii

carlossanlop committed:
Merge branch 'main' into bug/issue-112017-ii

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:
```diff
@@ -71,25 +70,43 @@ public static List<ZipGenericExtraField> ParseExtraField(ReadOnlySpan<byte> extr
+            if (!trailingExtraFieldData.IsEmpty)
             {
-                field.WriteBlock(stream);
+                stream.Write(trailingExtraFieldData);
```
Do you think there could be a case with no fields but with extra trailing data? Do we have a test for that?

edwardneal replied:
Thanks @carlossanlop. There can be, and I've just added test cases which cover this. The structure of an extra field is a two-byte tag ID, a two-byte data length, then a variable number of data bytes - so when we read four bytes of zeroes, we're actually reading a zero-length field with a tag ID of zero. The 0-3 bytes which follow this would be considered trailing data.

I've added tests which cover 3, 4, 7, 8 and 15 bytes of trailing data. This data should be interpreted by ZipArchive as 0, 1 or 2 well-formed zero-length fields, followed by 0 or 3 bytes of trailing data.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:
```diff
@@ -57,11 +56,11 @@ public static bool TryReadBlock(ReadOnlySpan<byte> bytes, out int bytesConsumed,
 
             field._data = bytes.Slice(FieldLocations.DynamicData, field._size).ToArray();
-            bytesConsumed += field._size;
+            bytesConsumed = field.Size + SizeOfHeader;
```
Good catch, we don't care about bytesConsumed if we return false. Better add the numbers only once.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:
```diff
@@ -71,25 +70,43 @@ public static List<ZipGenericExtraField> ParseExtraField(ReadOnlySpan<byte> extr
+            // fields in the central directory and local file headers are valid.
+            // We need to account for this because zipalign is part of the build process for Android applications, and failing to do this will result in
+            // .NET for Android generating corrupted .apk files and failing to build.
+            trailingExtraFieldData = extraFieldData[totalBytesConsumed..];
```
Thanks for the comment.

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -907,6 +907,382 @@ public static async Task ZipArchive_InvalidHuffmanData()
+
+                        Assert.Equal(10, bytesRead);
+
+                        Assert.Equal(new byte[] { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09 }, uncompressedBytes);
```
This expected array is used in the 3 new tests, consider making it a private field that the 3 tests can consume.

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -907,6 +907,382 @@ public static async Task ZipArchive_InvalidHuffmanData()
+        [InlineData(0x14)]
+        public void ZipArchive_InvalidExtraFieldData(byte validVersionToExtract)
+        {
+            byte[] invalidExtraFieldData = s_invalidExtraFieldData.AsSpan().ToArray();
```
This should be enough:
```suggestion
            byte[] invalidExtraFieldData = s_invalidExtraFieldData;
```

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -907,6 +907,382 @@ public static async Task ZipArchive_InvalidHuffmanData()
+
+                        Assert.Equal(4, bytesRead);
+
+                        Assert.Equal(new byte[] { 0x00, 0x01, 0x02, 0x03 }, uncompressedBytes);
```
This one too can be a shared private field.

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -907,6 +907,382 @@ public static async Task ZipArchive_InvalidHuffmanData()
+            invalidExtraFieldData[4] = validVersionToExtract;
+            invalidExtraFieldData[60] = validVersionToExtract;
+
+            using (MemoryStream updatedStream = new MemoryStream())
```
super nit: since these memory streams get disposed at the very end of the method can use the new using format and you can save a bunch of indentation in the next lines by removing the curly brackets that wrap all that code. The memory stream will be disposed at the end of the current context.

```suggestion
            using MemoryStream updatedStream = new MemoryStream();
```

carlossanlop reviewed (commented):
Thanks for the PR @edwardneal . Left some comments/questions for you to consider.

edwardneal committed:
Update test, adding test cases

Prove functionality where there are 3, 4, 7, 8 and 15 bytes of padding. This covers zero, one and multiple well-formed extra fields, with and without subsequent trailing data.

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -981,140 +982,191 @@ public static void ZipArchive_InvalidVersionToExtract()
-        [InlineData(0x00)]
-        [InlineData(0x14)]
-        public void ZipArchive_InvalidExtraFieldData(byte validVersionToExtract)
+        [InlineData(0, 3)]
+        [InlineData(20, 3)]
+        [InlineData(0, 4)]
+        [InlineData(20, 4)]
+        [InlineData(0, 7)]
+        [InlineData(20, 7)]
+        [InlineData(0, 8)]
+        [InlineData(20, 8)]
+        [InlineData(0, 15)]
+        [InlineData(20, 15)]
```
Not too important, I can add this later, probably in my PR since I would have to write an async version of this test.

But when you combine a bunch of inline data values, instead of adding a bunch of inline datas combining everything, you can do:

```suggestion
        [MemberData(nameof(Get_ZipArchive_InvalidExtraFieldData_Data))]
```
and then above the test you add this:
```cs
public static IEnumerable<object[]> Get_ZipArchive_InvalidExtraFieldData_Data()
{
    foreach (byte validVersionToExtract in new byte[] { 0, 20 })
    {
        foreach (ushort extraFieldDataLength in new ushort[] { 3, 4, 7, 8, 15 })
        {
            yield return new object[] { validVersionToExtract, extraFieldDataLength };
        }
    }
}
```

edwardneal replied:
It's simple enough to fix here - I've just switched it to using `MemberData`.

carlossanlop commented:
/azp run runtime-libraries-coreclr outerloop

carlossanlop commented:
Just in case, I ran the outerloop tests. If nothing concerning shows up, I'll approve and merge this.

edwardneal committed:
Refactored hardcoded inline test data

Converted to a clearer member which generates the combinations dynamically.

carlossanlop commented:
cc @jonathanpeppers this is the fix for the zipalign issue.

jonathanpeppers commented:
I will test this here:
* https://github.com/dotnet/android/pull/9840

The most recent Maestro PR didn't have this one yet:

* https://github.com/dotnet/android/pull/10008

I can try it on the next Maestro PR.

edwardneal commented:
Thanks @jonathanpeppers. I tested this PR earlier, and re-tested it with the most recent preview, using a new Android project with the settings:
```xml
<AndroidUseAapt2>True</AndroidUseAapt2>
<AndroidPackageFormat>apk</AndroidPackageFormat>
<AndroidSigningKeyStore>[keystore]</AndroidSigningKeyStore>
<AndroidSigningStorePass>[password]</AndroidSigningStorePass>
<AndroidSigningKeyPass>[password]</AndroidSigningKeyPass>
<AndroidKeyStore>true</AndroidKeyStore>
<AndroidSigningKeyAlias>apk</AndroidSigningKeyAlias>
<_AndroidUseLibZipSharp>false</_AndroidUseLibZipSharp>
```

I cleaned the project, then ran `dotnet publish -f net10.0-android -bl -v:diag`. The build failed and returned the output:
```
error ANDZA0000: 04-15 20:03:18.161  4356 15832 W zip     : WARNING: header mismatch
error ANDZA0000:                                                                    
error ANDZA0000: 04-15 20:03:18.164  4356 15832 W zip     : WARNING: header mismatch
error ANDZA0000:                                                                    
error ANDZA0000: 04-15 20:03:18.165  4356 15832 W zip     : WARNING: header mismatch
error ANDZA0000:                                                                    
error ANDZA0000: 04-15 20:03:18.165  4356 15832 W zip     : WARNING: header mismatch
error ANDZA0000:                                                                    
```

I've replaced .NET 10 Preview 3's System.IO.Compression.dll with the output from building dotnet/runtime, then cleaned the Android project again and re-ran `dotnet publish ...` with varying combinations of `AndroidUseAapt2 = True/False` and `AndroidPackageFormat = aab/apk`. The build completes without error, and when I use 7-Zip to test the archive, that test passes.

I couldn't force the zipalign step to fail, so had to test that independently - I just ran zipalign against an appropriate ZIP file, verified that it had the unusual padding in the extra data field, then used that as the initial baseline data for the new unit tests.

Are there any other tests I can run locally prior to the next Maestro PR?

jonathanpeppers commented:
This change hasn't shown up yet this morning:
* https://github.com/dotnet/android/pull/10041

I'll try again later, this is currently dotnet/runtime/main from ~April 11.

jonathanpeppers commented:
I just rebased this one, to test out this change:
* https://github.com/dotnet/android/pull/9840
